### ISSUE TITLE
Error writing to disk - not displayed

### ISSUE THREAD

**Issue by JakubFojtik** [mbusb/multibootusb#368]
After clicking "Write image to USB" Multibootusb instantly reports that image has been written, but it has not.
When i run the dd command manually I get _"Error opening output file: 32 The process cannot access the file because it is being used by another process"_.

I think that is just a common Windows error (restarting explorer.exe fixed it), probably not solvable in Multiboot, but it should not say it finished succesfully. See attached [multibootusb.log](https://github.com/mbusb/multibootusb/files/1989690/multibootusb.log):

```
00:07:41.655 root INFO Executing ==> C:\Program Files (x86)\multibootusb\data\tools\dd\dd.exe if=D:\moje\antergos-18.2-x86_64.iso of=\\.\E: bs=1M --progress
00:07:41.686 root INFO Error writing to disk...
00:07:41.686 root INFO ISO has been written to USB disk...
```

**Comment by shinji-s** [comment]
@JakubFojtik, were you able to boot off of the created usb stick after successfully writing to it?
 I can't. It seems to me "dd if=\\.\k: ..." writes to the partition, not the disk containing the partition.

**Comment by JakubFojtik** [comment]
@shinji-s I checked now and it does not boot: `isolinux.bin missing or corrupt`.
I believe it is because the \\.\PhysicalDriveX path stops working see #372.

You can verify that the USB boots if
`qemu-system-x86_64.exe -L . -boot c -hda //./e:`
is used instead of
`qemu-system-x86_64.exe -L . -boot c -hda //./PhysicalDriveX`

**Comment by mbusb** [comment]
@shiinji,
I too have similar issue. dd was successful but failed to boot. Writing ISO to //./PhysicalDriveX is resulting in Access Denied error. Neither `od=G:` option which selects whole disk instead of partition appears to work resulting in same access denied error. The only working solution for me is to flash it with `of=\\.\G:` which of course does not boot and neither it produces boot error. 

Does any one have the same issue?

**Comment by mbusb** [comment]
I have a thought. The `Access Denied` error could be related to some process/programs accessing USB silently in background. Similar error message appeared when the explorer.exe was open.

**Comment by JakubFojtik** [comment]
When you guys say "does not boot" do you mean only via Multibootusb's GUI? Because other ways work for me at least. Just the Multiboot's "qemu.exe \\.\PhysicalDrive1" does not work.

In my multibootusb.log dd uses "of=\\.\E:" and the resulting USB is bootable, just not from the GUI. It boots with "qemu.exe \\.\E:" when manually run in cmd.exe in qemu dir.

Do you run some antivirus software other than the default Defender? I do not, that may be a  factor.

**Comment by shinji-s** [comment]
@mbusb It seems to me Windows behaviour when writing to the sector zero has drastically changed recently. I wrote previously that I can run syslinux on an external HDD properly, but later I found out  that the the partition table was hosed (from which I could recover thanks to testdisk utility).
Could it be possible that Windows update released at the end of April has something to do with it?

FYI, rufus seems to have no trouble writing to a physical disk. It takes more elaborate steps to write to a physical disk than simply opening \\.\PhysicalDriveX.
https://github.com/pbatard/rufus/blob/c84ff4d121aec42967174270812ff851c3f2b8f9/src/format.c#L1628

**Comment by shinji-s** [comment]
Sorry, my assumption that writing to //./K: changes partition, not entire disk may be wrong.
I'll follow up with more info once I have some evidence to confirm or deny.

**Comment by shinji-s** [comment]
@JakubFojtik  My stick after dding onto \\.\K: does not boot even under a bare metal. (The same image/stick boots if I dd onto /dev/sdd on LInux).
Possibly your stick boots because the MBR on your stick chain boots to the partition boot sector or you dded EFI bootable OS image and you are using an EFI machine.
You may get different result if you wipe off the MBR before dding or use a BIOS machine.

I don't have any AV software running other than Microsoft provided one.

**Comment by shinji-s** [comment]
@mbusb
I wrote previously:

> It turned out that my Windows 10 installation is probably bugged. It exhibits same symptom as described in https://community.spiceworks.com/topic/1899241-diskpart-clean-fails-cannot-zero-sectors-on-disk-all-flash-drives
> and just running syslinux.exe on an usb flash drive fails though doing the same on a removable hdd works. Need to fix my computer before completing the fix of this issue.
> 

> It seems to me Windows behaviour when writing to the sector zero has drastically changed recently. I wrote previously that I can run syslinux on an external HDD properly, but later I found out that the the partition table was hosed (from which I could recover thanks to testdisk utility).

Sorry, that was a false alarm. I forgot that I had eanbled this part of the code in my workspace.
https://github.com/mbusb/multibootusb/blob/master/scripts/syslinux.py#L178
And quite likely that have lead to the strange Windows behavior and corruption of partition table.

Now what we should find is a way to write onto  \\\\.\\PhysicalDriveX.

**Comment by shinji-s** [comment]
@mbusb Thank you for the info. It appears that operation effectively unmounts the relevant volume(s).
Rufus explicitly unmounts volumes (and deletes mount points) before formatting a drive.
Here is an informative tip of theirs,  https://github.com/pbatard/rufus/issues/759

**Comment by shinji-s** [comment]
Is the change brought by the past few commits an introduction of dd_win_clean_usb() and config.imager_return?
I'm planning to move os dependent code back to osdriver.Widnows class.

**Comment by mbusb** [comment]
@shinji-s ,
Yes. Those were the changes. Make a pull request for merge.

**Comment by shinji-s** [comment]
I've brought my devel branch in sync with mbusb/devel and tried writing an iso file.
In the first attempt, dd exited immediately and the usb did not boot. The log is shown below.
(diskpart cleaned the disk successfully.)

> Physical Device Number is 2
> 
> Microsoft DiskPart バージョン 10.0.16299.15
> 
> Copyright (C) Microsoft Corporation.
> コンピューター: ESPRIMO-K553F
> 
> ディスク 2 が選択されました。(*1)
> 
> DiskPart はディスクを正常にクリーンな状態にしました。(*2)
> 
> DiskPart を終了しています... (*3)
> Executing ==> "C:\Users\shinj\Documents\src\multibootusb\data\tools\dd\dd.exe" if=C:\Users\shinj\Downloads\debian-9.4.0-amd64-netinst.iso of=\\.\PhysicalDrive2 bs=1M --progress
> No USB disk found...
> No USB device found...
> No USB device found...

In the 2nd attempt, where I placed a sleep of 5 seconds after diskpart/clean, dd was successful and the usb booted fine but it became impossible to diskpart/clean again until I get the disk reinitialized using rufus. It seems more testing and investigation is needed.

**Comment by shinji-s** [comment]
Got the following error during testing prior to sending a PR.
Simultaneously to getting the error, explorer showed warning that the filesystem can't be recognized and asked to insert a media. It seems \\.\PhysicalDriveX can go away in the middle of dding.

> Physical Device Number is 2
> 
> Microsoft DiskPart バージョン 10.0.16299.15
> 
> Copyright (C) Microsoft Corporation.
> コンピューター: ESPRIMO-K553F
> 
> ディスク 2 が選択されました。
> 
> DiskPart はディスクを正常にクリーンな状態にしました。
> 
> DiskPart を終了しています...
> Executing => ['C:\\Users\\shinj\\Documents\\src\\multibootusb\\data\\tools\\dd\\dd.exe', 'if=C:\\Users\\shinj\\Downloads\\openSUSE-Tumbleweed-NET-x86_64-Snapshot20180316-Media.iso', 'of=\\\\.\\physicaldrive2', 'bs=1M', '--progress']
> dd: rawwrite dd for windows version 0.6beta3.                            ]   6%
> dd: Written by John Newbigin <jn@it.swin.edu.au>
> dd: This program is covered by terms of the GPL Version 2.
> dd: 8M Error writing file: 2 指定されたファイルが見つかりません。
> dd: 9+0 records in
> dd: 8+0 records out
> Error writing iso image...

**Comment by mbusb** [comment]
I didn't experience such errors in yours and my commit. Tested with 4
different distros and all worked without crash. I think after changing the
filesystem, windows doesn't recognize the disk and asking user to format...

**Comment by shinji-s** [comment]
> I think after changing the filesystem, windows doesn't recognize the disk and asking user to format..

Yes. Windows asked for both inserting media and formatting the drive. (Each on different paritions?)
I got the above error in all three trials using the suse iso though the number of blocks successfully written differ. The error from debian-9.4.0-amd64-netinst.iso is more sporadic with higher success rate.

>

**Comment by CretaceousConstructor** [comment]
dd if=c.img of=\\.\E: bs=512 count=1
Error opening output file: 32 The process cannot access the file because it is being used by another process 
who can save me , I was driven crazy by this error

**Comment by CretaceousConstructor** [comment]
> dd if=c.img of=\.\E: bs=512 count=1
> Error opening output file: 32 The process cannot access the file because it is being used by another process
> who can save me , I was driven crazy by this error

I find that newer version of DD has a dd if=c:\my files\boot.bin od=z: bs=512 count=1 command which do the job.
